src/full/Agda/TypeChecking/Sort.hs

```diff
@@ -85,57 +86,66 @@ hasBiggerSort = void . inferUnivSort
 -- | Infer the sort of a pi type. If we can compute the sort straight away,
 --   return that. Otherwise, return @PiSort a s2@ and add a constraint to
 --   ensure we can compute the sort eventually.
-inferPiSort :: PureTCM m => Dom Type -> Abs Sort -> m Sort
+inferPiSort :: (PureTCM m,MonadConstraint m) => Dom Type -> Abs Sort -> m Sort
 inferPiSort a s2 = do
-  s1' <- reduce $ getSort a
-  s2' <- mapAbstraction a reduce s2
-  -- we do instantiateFull here to perhaps remove some (flexible)
-  -- dependencies of s2 on var 0, thus allowing piSort' to reduce
-  s2' <- instantiateFull s2'
-
+  s1' <- reduceB $ getSort a
+  s2' <- mapAbstraction a reduceB s2
   --Jesper, 2018-04-23: disabled PTS constraints for now,
   --this assumes that piSort can only be blocked by unsolved metas.
+  --Arthur Adjedj, 2023-02-27, Turned PTS back on,
+  --piSort can now be blocked by Leveluniv
+  let b' = unblockOnEither (getBlocker s1') (getBlocker $ unAbs s2')
+  case piSort' (unEl <$> a) (ignoreBlocking s1') (ignoreBlocking <$> s2') of
+    Right s -> return s
+    Left b -> do
+      addConstraint (unblockOnEither b b') $ HasPTSRule a (ignoreBlocking <$> s2')
+      return $ PiSort (unEl <$> a) (ignoreBlocking s1') (ignoreBlocking <$> s2')
 
-  --case piSort' s1 s2 of
-  --  Just s -> return s
-  --  Nothing -> do
-  --    addConstraint $ HasPTSRule s1 s2
-  --    return $ PiSort s1 s2
-
-  return $ piSort (unEl <$> a) s1' s2'
 
 -- | As @inferPiSort@, but for a nondependent function type.
 inferFunSort :: Sort -> Sort -> TCM Sort
-inferFunSort s1 s2 = funSort <$> reduce s1 <*> reduce s2
-
-ptsRule :: Dom Type -> Abs Sort -> Sort -> TCM ()
-ptsRule a b c = do
-  c' <- inferPiSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
-
--- | Non-dependent version of ptsRule
-ptsRule' :: Sort -> Sort -> Sort -> TCM ()
-ptsRule' a b c = do
-  c' <- inferFunSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
+inferFunSort s1 s2 = do
+  ss1 <- sortOf (Sort s1)
+  s1' <- reduceB s1
+  s2' <- reduceB s2
+  case funSort' (ignoreBlocking s1') (ignoreBlocking s2') of
+    Right s -> return s
+    Left b -> do
+      let b' = unblockOnEither (getBlocker s1') (getBlocker s2')
+      addConstraint (unblockOnEither b b') $ HasPTSRule (defaultDom (El (univSort s1) (Sort s1))) (mkAbs "_" $ ignoreBlocking s2')
```

Comment:
Here is probably a bug and the source of my pain with #6651.
`HasPTSRule` takes the domain and the sort of the codomain.
It looks like here it is invoked with the _sort_ of the domain rather than the domain itself.
In my example I get a constraint `HasPTSRule LockUniv LockUniv` which fails on `funSort Set1 LockUniv`.
Commenting out this line makes my example pass.
Seems like if `inferFunSort` wants to add a `HasPTSRule` constarint, it needs to get the domain as well as argument.
@arthur-adjedj CC @jespercockx

Reply:
This sounds right to me, we indeed need to have access to the full domain (and not just its sort) to create a `HasPTSRule` constraint. But perhaps it would be better to add a non-dependent version of the `HasPTSRule` constraint that does not require the domain (just as `funSort` is the non-dependent version of `piSort`).

Reply:
Ok, fixed in fcec24aeb4a782f1c2e5d0c009f1751ab968537e.